pkg/controller/helper/lookup_provider_account.go

```diff
@@ -0,0 +1,131 @@
+			return providerAccount, nil
+		}
+	}
+
+	// not found, return error
+	return nil, errors.New("LookupProviderAccount: no provider account found")
+}
+
+func providerAccountFromSecretReferenceSource(cl client.Client, ns string, providerAccountRef *corev1.LocalObjectReference, logger logr.Logger) (*ProviderAccount, error) {
+	if providerAccountRef != nil {
+		logger.Info("LookupProviderAccount", "ns", ns, "providerAccountRef", providerAccountRef)
+		secretSource := helper.NewSecretSource(cl, ns)
+		adminURLStr, err := secretSource.RequiredFieldValueFromRequiredSecret(providerAccountRef.Name, providerAccountSecretURLFieldName)
+		if err != nil {
+			return nil, fmt.Errorf("providerAccountFromSecretReferenceSource: %w", err)
+		}
+		token, err := secretSource.RequiredFieldValueFromRequiredSecret(providerAccountRef.Name, providerAccountSecretTokenFieldName)
+		if err != nil {
+			return nil, fmt.Errorf("providerAccountFromSecretReferenceSource: %w", err)
+		}
+
+		return &ProviderAccount{AdminURLStr: adminURLStr, Token: token}, nil
+	}
+
+	return nil, nil
+}
+
+func providerAccountFromDefaultSecretSource(cl client.Client, ns string, providerAccountRef *corev1.LocalObjectReference, logger logr.Logger) (*ProviderAccount, error) {
+	// if exists, fiels are required.
+	defaulSecret, err := helper.GetSecret(providerAccountDefaultSecretName, ns, cl)
+	if err == nil {
+		adminURLStr := helper.GetSecretDataValue(defaulSecret.Data, providerAccountSecretURLFieldName)
+		if adminURLStr == nil {
+			return nil, fmt.Errorf("providerAccountFromDefaultSecretSource: Secret field '%s' is required in secret '%s'", providerAccountSecretURLFieldName, defaulSecret.Name)
+		}
+		token := helper.GetSecretDataValue(defaulSecret.Data, providerAccountSecretTokenFieldName)
+		if token == nil {
+			return nil, fmt.Errorf("providerAccountFromDefaultSecretSource: Secret field '%s' is required in secret '%s'", providerAccountSecretTokenFieldName, defaulSecret.Name)
+		}
+
+		return &ProviderAccount{AdminURLStr: *adminURLStr, Token: *token}, nil
+	} else if err != nil && !apierrors.IsNotFound(err) {
+		return nil, fmt.Errorf("providerAccountFromDefaultSecretSource: %w", err)
+	}
+
+	return nil, nil
+}
+
+// Lookup default provider account for the 3scale deployment in the current namespace
+func providerAccountFromLocal3scaleSource(cl client.Client, ns string, providerAccountRef *corev1.LocalObjectReference, logger logr.Logger) (*ProviderAccount, error) {
+	// Read credentials and tenant url for default provider account of 3scale
+	listOps := []client.ListOption{client.InNamespace(ns)}
+	apimanagerList := &appsv1alpha1.APIManagerList{}
+	err := cl.List(context.TODO(), apimanagerList, listOps...)
+	if err != nil {
+		return nil, fmt.Errorf("providerAccountFromLocal3scaleSource: %w", err)
+	}
+
+	if len(apimanagerList.Items) == 0 {
+		return nil, nil
+	}
+
+	apimanager := apimanagerList.Items[0]
```

Comment:
Would it be worth it to check that there are no more than 1 APIManager? It shouldn't but I wonder whether checking that would be appropriate.

Reply:
enhancement!

If there are more than 1,  raise an error?